src/poseidon2/transcript.rs

```diff
@@ -88,61 +85,75 @@ impl<
         for chunk in field_els[add_to_last..].chunks(capasity_per_element) {
             let mut repr = <E::Fr as PrimeField>::Repr::default();
             for (i, el) in chunk.iter().enumerate() {
-                let mut value_repr = <E::Fr as PrimeField>::Repr::from(el.as_u64());
+                let mut value_repr = <E::Fr as PrimeField>::Repr::from(el.as_u64_reduced());
                 value_repr.shl((i * F::CHAR_BITS) as u32);
                 repr.add_nocarry(&value_repr);
             }
             self.buffer.push(E::Fr::from_repr(repr).unwrap());
         }
 
         self.last_filled = (self.last_filled + field_els.len()) % capasity_per_element;
+
+        self.available_challenges = vec![];
     }
 
     fn witness_merkle_tree_cap(&mut self, cap: &[Self::CompatibleCap]) {
         self.last_filled = 0;
         self.buffer.extend_from_slice(cap);
+
+        self.available_challenges = vec![];
     }
 
     fn get_challenge(&mut self) -> F {
         assert_eq!(self.sponge.filled, 0);
 
         if self.buffer.is_empty() {
             if self.available_challenges.len() > 0 {
-                let first_el = self.available_challenges.first().unwrap().into_repr();
-                let next_challenge;
-
-                if F::CHAR_BITS <= 64 {
-                    next_challenge = F::from_u64_with_reduction(first_el.as_ref()[self.challenges_taken]);
-                    self.challenges_taken += 1;
-                } else {
-                    todo!("Goldilocks has less than 64 bits per element");
-                }
-
-                if self.challenges_taken == Poseidon2Sponge::<E, F, M, RATE, WIDTH>::capasity_per_element() {
-                    self.available_challenges.drain(..1);
-                    self.challenges_taken = 0;
-                }
-
-                return next_challenge;
+                let first_el = self.available_challenges.first().unwrap().clone();
```

Comment:
consider making available_challenges a VecDeque 

this way, here you can simply do 'pop_front()'